### PR TITLE
[rlp] fix nested unbounded lists

### LINKED ISSUE #105: [RLP] Unexpected behavior when encoding nested lists with RLP
I have the following data structures defined as `Encodable` and `Decodable`:

```rust
#[derive(Debug, PartialEq, Eq)]
struct Inner(H256, H256);

impl Encodable for Inner {
	fn rlp_append(&self, s: &mut RlpStream) {
		s.begin_unbounded_list()
			.append(&self.0)
			.append(&self.1)
			.complete_unbounded_list();
	}
}

impl Decodable for Inner {
	fn decode(rlp: &Rlp) -> Result<Self, DecoderError> {
		Ok(Inner(rlp.val_at(0)?, rlp.val_at(1)?))
	}
}

#[derive(Debug, PartialEq, Eq)]
struct Nest(Vec<Inner>);

impl Encodable for Nest {
	fn rlp_append(&self, s: &mut RlpStream) {
		s.begin_list(1)
			.append_list(&self.0);
	}
}

impl Decodable for Nest {
	fn decode(rlp: &Rlp) -> Result<Self, DecoderError> {
		Ok(Nest(rlp.list_at(0)?))
	}
}
```

An encoding-decoding roundtrip of `Nest` fails like so:

Test: 
```rust
#[test]
fn test_nested_list_roundtrip() {
	let items = (0..8).map(|_| Inner(H256::random(), H256::random())).collect();
	let nest = Nest(items);

	let encoded = rlp::encode(&nest);
	let decoded = rlp::decode(&encoded).unwrap();

	assert_eq!(nest, decoded)
}
```
Error:
```
thread 'test_nested_list_roundtrip' panicked at 'assertion failed: `(left == right)`
  left: `Nest([Inner(0x347162d6b27b8f60eab7510e7c5f1a13cf8cf02128d4e0a75811bb6b21f924c1, 0x8a8681a07c0093db405e3462fcf77050f43b3ae8045494659d843e7d4c68c3dd), Inner(0x7b226f962317f5c23ed52c8920f22496014fba993aac53f1abefb38798a34aa0, 0x161de16522b077c1f7d05a369228c4ff2dceed19a5a70cc95dae7318ef273ac1), Inner(0x4c2c46c6f104296d554d2d46d7d87df9a7bd6096ea1ff3e46047378182507dc2, 0x633bbd4c3d50ef2cf2547659d4c5e77cbabd432f61461eb907352461872cf64f), Inner(0x0f07c3990ef3206c14bf5f7c24bb67fbae003aab039fbbcb11e01c9db8827376, 0x51e3c48e3be8e6ff7006f7d38d13c0f63371894df645a5321fbe2bda93fce29f), Inner(0x2c38e80703a49398c1972b43b8d3586e6fabbd8c8c9821ac02147c70fd790330, 0x1250dce35e0f998535255d7f9f442d945afe547adebbbaf71b38fa490171af2f), Inner(0xc8bf2de05f9601109e0de4a6a979ae7ccda4563d024b4f333a7187d05f0a7971, 0xee98d63d812605661a868dfccf508fc15cefcc4825c87c1883a43867f2c00a47), Inner(0x002c77234f46917656e6ef2d03d62baf066b8bb87cb4145a1a485f86fe626e53, 0x031f58ed89b8b8b7660772a2a7c8669e9ee6298d09f7edc09e0a7f0a554e7436), Inner(0x85425be5867ca860f1f568a08f0d13dd8a70bdb8ade9db1d0cc8e65899a4a6d9, 0x17d50a0baf5975038eaec5667cb4df6f1c49c5bf34c0897ed59559972a3a5554)])`,
 right: `Nest([Inner(0x347162d6b27b8f60eab7510e7c5f1a13cf8cf02128d4e0a75811bb6b21f924c1, 0x8a8681a07c0093db405e3462fcf77050f43b3ae8045494659d843e7d4c68c3dd), Inner(0x7b226f962317f5c23ed52c8920f22496014fba993aac53f1abefb38798a34aa0, 0x161de16522b077c1f7d05a369228c4ff2dceed19a5a70cc95dae7318ef273ac1), Inner(0x4c2c46c6f104296d554d2d46d7d87df9a7bd6096ea1ff3e46047378182507dc2, 0x633bbd4c3d50ef2cf2547659d4c5e77cbabd432f61461eb907352461872cf64f), Inner(0x0f07c3990ef3206c14bf5f7c24bb67fbae003aab039fbbcb11e01c9db8827376, 0x51e3c48e3be8e6ff7006f7d38d13c0f63371894df645a5321fbe2bda93fce29f)])`', tests/tests.rs:546:2
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

It seems like `rlp` crate only encodes half of the nested list (4 items of `nest.0` in this case instead of 8) when the full list is expected.

Changing initial amount of objects in `nest.0` to any other number yields the same result - only half of the vec is encoded.

### PR DESCRIPTION
**Pull request by ordian**
Fixes #105.

### CODE REVIEW

File: rlp/src/stream.rs [hunk 1]
```diff
@@ -335,6 +334,7 @@ impl RlpStream {
 		let len = self.buffer.len() - list.position;
 		self.encoder().insert_list_payload(len, list.position);
 		self.note_appended(1);
+		self.finished_list = true;
```
Comment (ordian) [inline rlp/src/stream.rs]: the fix is here

File: rlp/src/stream.rs [hunk 2]
```diff
@@ -326,7 +325,7 @@ impl RlpStream {
 		BasicEncoder::new(self)
 	}
 
-	/// Finalize current ubnbound list. Panics if no unbounded list has been opened.
+	/// Finalize current unbounded list. Panics if no unbounded list has been opened.
 	pub fn complete_unbounded_list(&mut self) {
```
Comment (dvdplm) [inline rlp/src/stream.rs]: Unrelated, but maybe this should be called `finalize_unbounded_list()`?

Comment (ordian) [reply rlp/src/stream.rs]: Both sounds fine to me, but I'm not a native speaker. Also, I'd like to avoid breaking changes.

Comment (dvdplm) [reply rlp/src/stream.rs]: "complete" can be both the imperative form of the verb as in "complete your homework before dinner!" or be and adjective "a complete list of my belongings". 
"finalize" does not have this ambiguity.

In this context the ambiguity makes it unclear if the method returns an unbounded list of something that is complete, or that it takes an unbounded list and completes it.

We can add a copy named `finalize_unbounded_list` and deprecate this one. In another PR.

Comment (ordian) [reply rlp/src/stream.rs]: opened #204

**Pull request merged**